File: Striver/DP/Q51.py
# burst balloons

# recursive solution
def maxCoins(i,j,nums):
    if i > j: return 0
    mx = float('-inf')

    for k in range(i,j+1):
        cost = nums[i-1]*nums[k]*nums[j+1] + maxCoins(i,k-1,nums) + maxCoins(k+1,j,nums)
        mx = max(mx,cost)

    return mx

# memoization solution
def maxCoinsMemo(i,j,nums,memo):
    if i > j: return 0
    if memo[i][j] != -1: return memo[i][j]
    mx = float('-inf')

    for k in range(i,j+1):
        cost = nums[i-1]*nums[k]*nums[j+1] + maxCoinsMemo(i,k-1,nums,memo) + maxCoinsMemo(k+1,j,nums,memo)
        mx = max(mx,cost)

    memo[i][j] = mx
    return mx

# tabulation solution
def maxCoinsTab(nums):
    n = len(nums)
    dp = [[0]*(n+2) for _ in range(n+2)]
    nums = [1] + nums + [1]

    for i in range(n, 0, -1):
        for j in range(1, n+1):
            if i > j: continue

            for k in range(i, j+1):
                dp[i][j] = max(dp[i][j], nums[i-1]*nums[k]*nums[j+1] + dp[i][k-1] + dp[k+1][j])
    
    return dp[1][n]


nums = [3,1,5,8]
n = len(nums)
memo = [[-1]*(n+2) for _ in range(n+2)]
newNums = [1] + nums + [1]
print(maxCoins(1,n,newNums))
print(maxCoinsMemo(1,n,newNums,memo))
print(maxCoinsTab(nums))

File: Striver/DP/Q51.java
package Striver.DP;

public class Q51{
    // burst balloons

    // recursive solution - Time: O(3^n) Space: O(n)
    public int maxCoins(int i, int j, int[] nums) {
        if(i > j) return 0;
        int max = Integer.MIN_VALUE;

        for(int ind = i; ind <= j; ind++){
            int cost = nums[i-1] * nums[ind] * nums[j+1] + maxCoins(i, ind-1, nums) + maxCoins(ind+1, j, nums);
            max = Math.max(max, cost);
        }

        return max;
    }

    // memoization - Time: O(n^3) Space: O(n^2)
    public int maxCoinsMemo(int i, int j, int[] nums, int[][] dp) {
        if(i > j) return 0;
        if(dp[i][j] != 0) return dp[i][j];
        int max = Integer.MIN_VALUE;

        for(int ind = i; ind <= j; ind++){
            int cost = nums[i-1] * nums[ind] * nums[j+1] + maxCoinsMemo(i, ind-1, nums, dp) + maxCoinsMemo(ind+1, j, nums, dp);
            max = Math.max(max, cost);
        }

        return dp[i][j] = max;
    }

    // tabulation - Time: O(n^3) Space: O(n^2)
    public int maxCoinsTab(int[] nums){
        int n = nums.length;
        int[][] dp = new int[n+2][n+2];

        int[] arr = new int[n+2];
        arr[0] = arr[n+1] = 1;
        for(int i = 1; i <= n; i++) arr[i] = nums[i-1];

        for(int i = n; i >= 1; i--){
            for(int j = i; j <= n; j++){

                if(i > j) continue;

                for(int k = i; k <= j; k++){
                    int cost = arr[i-1] * arr[k] * arr[j+1] + dp[i][k-1] + dp[k+1][j];
                    dp[i][j] = Math.max(dp[i][j], cost);
                }
            }
        }

        return dp[1][n];
    }

    public static void main(String[] args) {
        int[] nums = {3,1,5,8};
        Q51 q = new Q51();
        int[] newNums = new int[nums.length+2];
        newNums[0] = newNums[nums.length+1] = 1;
        for(int i = 1; i <= nums.length; i++) newNums[i] = nums[i-1];
        System.out.println(q.maxCoins(1, nums.length, newNums));
        System.out.println(q.maxCoinsMemo(1, nums.length, newNums, new int[nums.length+2][nums.length+2]));
        System.out.println(q.maxCoinsTab(nums));
    }
}